pedsnetdcc/parallel_db_exec.py

```diff
@@ -0,0 +1,138 @@
+from multiprocessing import Process, Queue
+import psycopg2
+from queue import Empty
+import threading
+
+from pedsnetdcc.dict_logging import DictQueueHandler
+
+logger = logging.getLogger(__name__)
+
+
+def parallel_db_exec(conn_str, sqls, count=None):
+    """Executes a dict of sqls in parallel.
+
+    `conn_str` is the database connection string to connect with.
+    `sqls` is a dict of arbitrary names to sql statements that should be
+    executed.
+    `count` is the number of results to fetch from the db cursor.
+
+    The `sqls` dict has the sql statement values replaced with the results
+    of that statement. The result format is a dict with two entries: `data`
+    which holds the single iterable result row if count is 1 or the iterable
+    of iterable results rows otherwise and `field_names` which holds an
+    iterable of result row field names in the same order as the result row(s)
+    in `data`.
+    """
+
+    resq = Queue()
+    logq = Queue()
+    workers = []
+
+    # Start the worker processes.
+    for name, sql in sqls.items():
+        wp = Process(target=db_exec, name=name, args=(conn_str, sql, count,
+                                                      name, resq, logq))
+        workers.append(wp)
+        wp.start()
+
+    # Start the logging thread to receive logs from the workers
+    logp = threading.Thread(target=logger_thread, args=(logq,))
+    logp.start()
+
+    # Wait for all the workers to finish.
+    for wp in workers:
+        wp.join()
+
+    # End the logging thread.
+    logq.put(None)
+    logp.join()
+
+    # Collect the results.
+    while True:
+        try:
+            result = resq.get_nowait()
+            sqls[result['name']] = result['output']
+        except Empty:
+            break
+
+    return sqls
+
+
+def logger_thread(q):
+    """Passes log records from a queue on to the logger.
+
+    Stops when None is retrieved from the queue.
+    """
+
+    while True:
+        record = q.get()
+        if record is None:
+            break
+        if logger.isEnabledFor(record.levelno):
+            logger.handle(record)
+
+
+def db_exec(conn_str, sql, count=None, name=None, resq=None, logq=None):
+    """Executes a sql statement against the database.
+
+    `conn_str` is the connection string to use.
+    `sql` is the statement to execute.
+    `count` is the number of records to return, all is the default.
+    `name` is an arbitrary name for the statement.
+    `resq` is the queue to pass results to.
+    `logq` is the queue to log to.
+
+    The return value is a dict with two entries: `name` which holds the passed
+    name and `output` which is a dict also with two entries: `data` which holds
+    the single result row iterable if count is 1 or the iterable of result row
+    iterables otherwise and `field_names` which is an iterable of result row
+    field names in the same order as the result row iterable(s).
+    """
+
+    # Configure logging to the queue if it is passed.
```

Comment:
My understanding is that if not `logq`, then there will be a run-time error because the conditional assignment to `logger` turns it into a local, so all references to `logger` in that function now refer to the local variable. So inside `db_exec`, instead of `logger`, you could use `loc_log`, and in the case where not `logq`, then say `loc_log = logger`.